components/tikv_alloc/src/lib.rs

```diff
@@ -82,19 +82,18 @@
 #[macro_use]
 extern crate log;
 
-#[cfg(not(all(unix, not(fuzzing), feature = "jemalloc")))]
+#[cfg(all(unix, not(fuzzing), not(feature = "jemalloc")))]
 mod default;
 
 pub type AllocStats = Vec<(&'static str, usize)>;
 
-// Allocators
 #[cfg(all(unix, not(fuzzing), feature = "jemalloc"))]
 #[path = "jemalloc.rs"]
 mod imp;
 #[cfg(all(unix, not(fuzzing), feature = "tcmalloc"))]
 #[path = "tcmalloc.rs"]
 mod imp;
-#[cfg(not(all(unix, not(fuzzing), any(feature = "jemalloc", feature = "tcmalloc"))))]
+#[cfg(all(unix, not(fuzzing), feature = "system-alloc"))]
```

Comment:
On Windows and WASM none of the `mod imp;` will be present, causing compilation error. But TiKV won't work on those platforms anyway, so it doesn't really matter??

Reply:
@siddontang @brson PTAL

Reply:
@BusyJay can you test it on Windows?

Reply:
I think the logic here needs to change a bit. The fuzzing case needs to work automatically, which I think it does not here, and people do build tikv on windows, plus we don't need to be _hostile_ to windows.

I do love how clean this logic is though.

The easiest way to accommodate the fuzzer and windows may be to just add the following:

```
#[cfg(any(fuzzing), windows)]
#[path = "system.rs"
mod imp;
```

If we add that, it probably enough to land.

There are probably still more things to improve though. For example the case where nobody picks an allocator will result in a mysterious compiler error. It would be better in that case to use `compile_error!` to emit something readable.

I also wish there was a `compile_warn!` macro so we could indicate in the fuzzer/windows case that the allocator is being "downgraded".

Reply:
For the wasm case, I don't think it's much of a concern right now unless we know of somebody trying to compile to wasm. There's are going to be significant other challenges with that port.

Reply:
The problem with my suggestion above is that on windows, users _don't_ need to pick an allocator - system will be chosen automatically.

Reply:
FWIW [this](https://github.com/brson/megamalloc/blob/master/src/lib.rs#L14) is the logic I've set up in my external version of this crate, with the [features](https://github.com/brson/megamalloc/blob/master/Cargo.toml#L7) organized in a way that more easily indicates that _some allocator_ was picked, making it easier to add `compile_err!` messages.

I haven't though considered the windows case. And neither tikv_alloc nor megamalloc deal cleanly with the case that _two_ allocators are selected.

Don't copy that crate here though. What we've got here is good enough for now, with the addition of the fuzzer and windows conditions.

Organizing these features in a foolproof way is quite a difficult puzzle!

Reply:
I can do a build on windows once the support is added in.